### ISSUE TITLE
Include lib in lizard-client

### ISSUE THREAD

**Issue by ernstkui** [nens/lizard-javascript-api#6]
*TL;DR* after doing some of the todos of this repo it will be 14 days of work to implement it in Lizard-client.

https://github.com/nens/lizard-client/pull/750 Replaces lizard-clients `timeseries` for `selections`, a superset of `timesereries` including `rasters`. As a result the lizard-client's ui may threat raster data the same as timeseries data: they have a color, and a position in the dashboards which can be serialized in a favourite. Hooray!

**Problem.** https://github.com/nens/lizard-client/pull/750 does not yet take into account syncing server data with _selections_ and it is not yet possible to make a selection of an eventseries. This library was started because I figured it would be an overhaul of lizard-client's global state, data-service, raster-service and timeseries-service in an aging app/dialect/framework while solving generic problems for every lizard client, in-house or external (kpi, waterlabel, HKV, Tijhuis, Hydrologic, Digitale Delta). A seperate library  creates benefits:

- It removes a lot of boilerplate for new javascript clients.
- It acts as an explicit convention for clients to interact with the Lizard API.
- It makes it a lot easier to build apps on top of Lizard by developers who have not built Lizard or spend months physically in the same room as the people who build Lizard.

But the original goal was to make  https://github.com/nens/lizard-client/pull/750 work. What needs to be done to make https://github.com/nens/lizard-client/pull/750 work in this library:

- [ ] Remove all timeseries of asset when removing asset.
- [ ] Include eventseries.
- [ ] Remove all intersections when raster/eventeries/timeseries are removed.
- [ ] Check user input. Currently an action payload is passed all the way through without ever checking what is inside. Throw errors.
- [ ] Throw errors and call error callbacks when request goes bad or response is malformed.
- [ ] Cancel requests when no longer relevant.

Following we need to find map the way this lib handles state, data, actions  and the way Lizard-client expects it. Lizard has a `State` object which keeps track of all ui state without server side data. Items in `State` match items in `Data-service` and `Timeseries-Service` asynchronously. Lizard js API follows [Redux](http://redux.js.org/docs/introduction/CoreConcepts.html) (it is a redux store) so it has a single source of truth.

**Assets in lizard-client versus Lizard-JavaScript-API**
Adding an asset in lc:
```js
State.assets.addAsset('entity$id');
```
versus lizard-javascript api:
```js
store.dispatch(addAsset(entity, id));
```
Gives lc:
```js
> State.assets;
< ['entity$id']
```
Lizard-js-api:
```js
> store.getState().assets;
< { 'entity$id': {entity: entity, id: id} }
```
After response in lc:
```js
> DataService.assets;
< [{entity: entity, id: id, ...reponse}];
```
After response in Lizard-js-api:
```js
> store.getState().assets;
< { 'entity$id': {entity: entity, id: id, ...data} }
```
This would require a fairly simple rewrite in lizard-client. 

- Some coupling between Angular's scope variables in omnibox and the Redux store. (1)
- Dispatch actions for search and map clicks instead of `addAsset`. (1)

**Layers of type raster versus rasters**
Lizard-client (LC) has layers: a collection of metadata for wmslayers, assetgroup(s), scenarios and rasters. Lizard javascript API (LAPI) only has rasters in objects. LC rasters can be on or off, LAPI has no such notion. I suggest to keep LC layers in place and not to use LAPI rasters in LC. This can be done because intersections in LAPI are fully seperate from rasters. (0)
 
**Selections versus Intersections**
LC stores timeseries metadata in `DataService.assets[i].timeseries`. LAPI has it in `store.getState().timeseries` with a possible reference to an asset (it is or should also be possible to store timeseries metadata without assets). An intersection is like a selection in that it is a slice of a raster or timeseries through time or space. The dashboard in LC would show intersections of all timeseries, eventseries and rasters for one timestate and all locations of assets and geometries, grouped by asset.  

In LC time is singular and always `State.temporal`, space is a list in `State.geometries`. In LAPI every intersection has its own space or time @roelvandenberg is proposing to normalize this in which case you would have multiple spaces and times which relate to one or more intersections.

With LAPI, LC would have to add intersections for every geometry and timeseries. And when `State.temporal` changes it will have to update all intersections with a new time interval.

I guess it will require less code in LC but it is hairy. (5)

LC stores position in dashboard and color on intersection. I think this is lizard-client specific and should not be part of LAPI, therefore it needs its own reducer to keep act on certain actions or it needs to define its own entities with references to LAPI entities. (3)

**Favourites**
LC serializes its `State` object which is stored on the server. In Redux it is possible to initialize a store with JSON and it is possible to dispatch actions to modify the state. 

A favourite state object with LAPI is simply `store.getState()`. This contains everything. When the server sends such an object to the browser it needs to filter out everything the user is not authorised for. After that it is a fully hydrated state object. Upside, restoring a state will not trigger a waterfall of requests.

Frontend (3). Backend (3).

Downside, favourites will increase in size, and it adds a lot of redundancy, storing in JSON objects which are also stored normalized in the database. In the future we could work around this by using LAPI on the backend to recreate a favourite object with a subset of data.

**NOTE**
Once this library is implemented in other apps it will be significantly harder to make changes to it. Consider implementing [selectors](http://redux.js.org/docs/recipes/ComputingDerivedData.html) first. By changing state and selectors we might be able to make bigger changes while staying backwards compatible for longer.
